stage0/src/dice_attestation.rs

```diff
@@ -44,64 +53,96 @@ fn generate_stage1_certificate(
+            Value::Bytes(measurements.cmdline_measurement.into()),
         ),
         (
             ClaimName::PrivateUse(SETUP_DATA_MEASUREMENT_ID),
-            Value::Bytes(Vec::from(measurements.setup_data_measurement)),
+            Value::Bytes(measurements.setup_data_measurement.into()),
         ),
         (
             ClaimName::PrivateUse(INITRD_MEASUREMENT_ID),
-            Value::Bytes(Vec::from(measurements.ram_disk_measurement)),
+            Value::Bytes(measurements.ram_disk_measurement.into()),
         ),
         (
             ClaimName::PrivateUse(MEMORY_MAP_MEASUREMENT_ID),
-            Value::Bytes(Vec::from(measurements.memory_map_measurement)),
+            Value::Bytes(measurements.memory_map_measurement.into()),
         ),
         (
             ClaimName::PrivateUse(ACPI_MEASUREMENT_ID),
-            Value::Bytes(Vec::from(measurements.acpi_measurement)),
+            Value::Bytes(measurements.acpi_measurement.into()),
         ),
     ];
     generate_eca_certificate(stage0_eca_key, stage0_cert_issuer, additional_claims)
+        .expect("couldn't generate ECA certificate")
 }
 
-/// Generates signed attestation for the 'measurements' of all Stage 1 components.
-pub fn generate_stage1_attestation(measurements: &Measurements) {
-    // Generate Stage0 keys. This key will be used to sign Stage1
-    // measurement+config and the stage1_ca_key.
-    let (stage0_eca_key, stage0_eca_verifying_key) = generate_ecdsa_key_pair();
+/// Generates attestation evidence for the 'measurements' of all Stage 1 components.
+pub fn generate_dice_data(measurements: &Measurements) -> &'static Stage0DiceData {
+    let result = Box::leak(Box::new_in(
+        Stage0DiceData::new_zeroed(),
+        &crate::BOOT_ALLOC,
+    ));
 
-    // Call generate_attestation_report() to get 'stage0_ca_verifying_key' added to attestation
-    // report.
-    log::debug!(
-        "Stage0 Verification key: {}",
-        hex::encode(stage0_eca_verifying_key.to_encoded_point(false).as_bytes())
-    );
+    // Generate ECA Stage0 key pair. This key will be used to sign Stage1 ECA certificate.
+    let (stage0_eca_key, stage0_eca_verifying_key) = generate_ecdsa_key_pair();
 
-    // Generate Stage1 CWT.
-    let stage1_eca = generate_stage1_certificate(
+    let (stage1_eca_cert, stage1_eca_signing_key) = generate_stage1_certificate(
         &stage0_eca_key,
         hex::encode(derive_public_key_id(&stage0_eca_verifying_key)),
         measurements,
     );
-    if stage1_eca.is_ok() {
-        // Call code that transmits the following to Stage1.
-        // 1. stage0_eca_verifying_key
-        // 2. stage1_eca.0 (the Stage 1 evidence)
-        // 3. stage1_eca.1 (the Stage 1 ECA private key)
-        log::debug!("Signing was successful..");
-        return;
-    }
-    log::debug!("Error in signing: {}", stage1_eca.unwrap_err())
+
+    let stage0_eca_verifying_key = public_key_to_cose_key(&stage0_eca_verifying_key)
```

Comment:
I don't really like the fact that we're allocating a bunch of vectors only to copy the data out of there int a slice, but I don't think there's a better way to do it either :(

Reply:
Agreed, but the coset APIs (and the underlying ciborium APIs) use vecs and the hashes/crypto stuff mostly return arrays.